string/string.h

```diff
@@ -0,0 +1,273 @@
+        }
+
+        this->string[position] = character;
+        this->strlen++;
+    }
+
+    /**
+     * Removes character at position specified
+     * @param position to remove at
+     */
+    void remove(int position) {
+        if (position > strlen) return;
+
+        for (int i = position; i < strlen + 1; i++) {
+            this->string[i] = this->string[i + 1];
+        }
+        this->strlen--;
+    }
+
+    /**
+     * Wipes the entire string buffer
+     */
+    void clear() {
+        int i = 0;
+        while (this->strlen > 0) {
+            string[i++] = NULL_TERMINATOR;
+            this->strlen--;
+        }
+    }
+
+    /**
+     * Recount the length of the string in case
+     * null gets randomly added to a string
+     *
+     * @return New Length or -1 if unchanged
+     */
+    int recount() {
+        int len_count = 0;
+
+        while (this->string[(len_count++) + 1] != NULL_TERMINATOR);
+
+        if (len_count != this->strlen) {
+            this->strlen = len_count;
+
+            return this->strlen;
+        }
+
+        return -1;
+    }
+
+    /**************************************
+     * Iterators
+     **************************************/
+    // TODO: Could be nice. Maybe add later after getting base functionality down.
+
+    /**************************************
+     * Capacity
+     **************************************/
+
+    /**
+     * Check if string is empty
+     * @return
+     */
+    bool empty() const {
+        return this->strlen == 0;
+    }
+
+    /**
+     * Check if string fills entire buffer
+     * @return
+     */
+    bool full() const {
+        return this->strlen >= size;
+    }
+
+    /**
+     * Get the max size of the buffer
+     * @return
+     */
+    size_t max_size() const {
+        return size;
+    }
+
+    /**
+     * Get the remaining available space in the buffer
+     * @return
+     */
+    size_t available() const {
+        return size - strlen - 1;
+    }
+
+    /**************************************
+     * C API Interfacing
+     **************************************/
+
+    /**
+     * Checks to see if the string buffer is full
+     *
+     * @return boolean on if there buffer is filled
+     */
+    bool is_full() const {
+        return this->strlen >= size;
+    }
+
+    /**
+     * Checks if there is extra space in the buffer
+     * @return boolean on if there's extra space
+     */
+    bool is_trunc() const {
+        return this->strlen < size;
+    }
+
+    /**
+     * Get the first terminator
+     * @return
+     */
+    char *get_end() const {
+        return this->string + len();
+    }
+
+protected:
+    String(char *buffer, size_t size): size(size), string(buffer) {}
+};
+
+namespace str {
+    template<const size_t SIZE>
+    class String : public ::String {
+    public:
+        /**
+         * String constructor
+         * @param buffer
+         */
+        explicit String(char *buffer) : ::String(buffer, SIZE) {};
```

Comment:
this constructor should take no arguments, 'buffer' should be a private member variable which is a SIZE big array

Reply:
you probably also want to zero out the buffer either here or in the protected constructor for ::String